### PR TITLE
Conda environments

### PR DESCRIPTION
**Pull request by isuruf**
This PR adds support for conda environments.

With conda environments, you can now install packages in conflict with Python 2.7 (eg. packages linking against msvc2015_runtime are in conflict with Python 2.7 and there was no way to install them before.)

You can also use a systemwide conda environment as well. 
```
conda create -n julia python=3.5
source activate julia
export CONDA_JL_HOME=/home/isuru/miniconda3/envs/julia
julia -e `Pkg.build("Conda")
```

### CODE REVIEW

**Comment by isuruf** [comment]
Closing and opening to trigger CI

**Comment by isuruf** [comment]
Tests pass now. Appveyor failure is in master also.
@Luthaf can you review?

**Comment by Luthaf** [comment]
Appveyor failure on [master](https://ci.appveyor.com/project/StevenGJohnson/conda-jl/build/1.0.2/job/tk2pjla1glc76pr1) do not have the same origin as [this PR](https://ci.appveyor.com/project/StevenGJohnson/conda-jl/build/1.0.4/job/97oc618dw5veox45).

File: deps/build.jl [hunk 1]
```diff
@@ -0,0 +1,4 @@
+default_dir = get(ENV, "CONDA_JL_HOME", abspath(dirname(@__FILE__), "usr"))
```
Comment (Luthaf) [inline deps/build.jl]: Why use an environment variable to specify the conda root path? I don't see the use case here.

File: src/Conda.jl [hunk 2]
```diff
@@ -34,43 +34,95 @@ module Conda
...
 import Compat.String
 using JSON
+include("../deps/deps.jl")
 
-"Prefix for installation of all the packages."
-const PREFIX = abspath(dirname(@__FILE__), "..", "deps", "usr")
+if !isdir(default_dir)
+    # Ensure default_dir exists
+    mkpath(default_dir)
+end
```
Comment (Luthaf) [inline src/Conda.jl]: Could this be moved to the `_install_conda` function?

File: src/Conda.jl [hunk 3]
```diff
@@ -34,43 +34,95 @@ module Conda
...
+
+type Environment
```
Comment (Luthaf) [inline src/Conda.jl]: We might want to have this as an immutable type.

File: src/Conda.jl [hunk 4]
```diff
@@ -34,43 +34,95 @@ module Conda
...
+    return is_windows() ? prefix(env) : bin_dir(env)
+end
+const PYTHONDIR = python_dir(RootEnv)
 
-"Path to the `conda` binary"
-const conda = joinpath(SCRIPTDIR, "conda")
+function conda_bin(env::Environment)
+    if is_windows()
+        p = script_dir(env)
+        return isfile(joinpath(p, "conda.bat")) ? joinpath(p, "conda.bat") : joinpath(p, "conda.exe")
```
Comment (Luthaf) [inline src/Conda.jl]: Why is this necessary? Just using `joinpath(script_dir(env), "conda")` should (and did previously) work on Windows and Unix, because Windows will look for any file with any of the `%PATHEXT%` extension.

File: src/Conda.jl [hunk 5]
```diff
@@ -34,43 +34,95 @@ module Conda
...
 
 """
 Use a cleaned up environment for the command `cmd`.
 
 Any environment variable starting by CONDA will interact with the run.
 """
-function _set_conda_env(cmd)
-    env = copy(ENV)
+function _set_conda_env(cmd, env::Environment=RootEnv)
+    env_hash = copy(ENV)
```
Comment (Luthaf) [inline src/Conda.jl]: This name is misleading (at least to me =)). I Though this was the result of `hash(env)`. How about `env_var` or `variables` or `env_variable`?

File: src/Conda.jl [hunk 6]
```diff
@@ -137,34 +187,40 @@ function _install_conda(force=false)
...
+        run(_set_conda_env(`$(conda_bin(RootEnv)) create -y -p $(prefix(env)) zlib`, RootEnv))
     end
 end
 
 "Install a new package."
-function add(pkg::AbstractString)
-    _install_conda()
-    run(_set_conda_env(`$conda install -y $pkg`))
+function add(pkg::AbstractString, env::Environment=RootEnv)
+    _install_conda(false, env)
```
Comment (Luthaf) [inline src/Conda.jl]: The `false` parameter should not be needed with the default parameters. You can inverse the parameter order in `_install_conda` to be `_install_conda(env::Environment, force=false)`. This way you do not even need the default value for `env`, which will be provided by all the callers.

Comment (isuruf) [inline src/Conda.jl]: I was keeping that for backwards compatibility. Since it's an internal function, I guess that's not an issue. I'll change it.

File: src/Conda.jl (hunk 4 above)
Comment (isuruf) [inline src/Conda.jl]: In an environment this is bit different because there are two files, `conda` and `conda.bat`. First one is a bash script for MSYS2. Using `conda` would pick up this file but it is not an executable.

File: deps/build.jl (hunk 1 above)
Comment (isuruf) [inline deps/build.jl]: This would enable the use of an existing conda environment, so that you can use an environment like `julia` in your system-wide environment.

**Comment by isuruf** [comment]
I checked master here, https://ci.appveyor.com/project/isuruf/conda-jl/build/1.0.6/job/4qbq0g3crpws54o6 and it failed exactly the same way as this PR.

File: src/Conda.jl (hunk 2 above)
Comment (isuruf) [inline src/Conda.jl]: When creating an environment using a full path, I check that the dir exists. I can remove that check and move this to `_install_conda`

File: deps/build.jl [hunk 7]
```diff
@@ -0,0 +1,4 @@
+default_dir = get(ENV, "CONDA_JL_HOME", abspath(joinpath(dirname(@__FILE__), "usr")))
```
Comment (stevengj) [inline deps/build.jl]: It should remember the environment variable so that you only need to set it once, and `Pkg.update()` won't "forget" it.  The way I do this in PyCall is to store it in a file.   e.g. do:
```jl
default_dir = get(ENV, "CONDA_JL_HOME", isfile("CONDA_HOME") ? readchomp("CONDA_HOME") : abspath(joinpath(dirname(@__FILE__), "usr")))
```
and then, at the end of the build script (i.e. after a successful build) do:
```jl
open("CONDA_HOME") do io
    println(io, default_dir)
end
```
Alternatively, you could use the `deps.jl` file that you are already writing, i.e. do
```jl
if haskey(ENV, "CONDA_JL_HOME")
    default_dir = ENV["CONDA_JL_HOME"]
elseif isfile("deps.jl")
    include("deps.jl")
else
    default_dir = abspath(dirname(@__FILE__), "usr")
end
```

File: src/Conda.jl [hunk 8]
```diff
@@ -34,43 +34,95 @@ module Conda
 using Compat
 import Compat.String
 using JSON
+include("../deps/deps.jl")
```
Comment (stevengj) [inline src/Conda.jl]:
```jl
if isfile("../deps/deps.jl")
    include("../deps/deps.jl")
else
   error("Conda is not properly configured.  Run Pkg.build(\"Conda\") before importing the Conda module.")
end
```

File: src/Conda.jl [hunk 9]
```diff
@@ -34,43 +34,95 @@ module Conda
...
 
-"Prefix for installation of all the packages."
-const PREFIX = abspath(dirname(@__FILE__), "..", "deps", "usr")
+if !isdir(default_dir)
+    # Ensure default_dir exists
+    mkpath(default_dir)
+end
+
+immutable Environment
+    path::ASCIIString
```
Comment (stevengj) [inline src/Conda.jl]: Compat.ASCIIString

File: src/Conda.jl [hunk 10]
```diff
@@ -34,43 +34,95 @@ module Conda
...
+include("../deps/deps.jl")
 
-"Prefix for installation of all the packages."
-const PREFIX = abspath(dirname(@__FILE__), "..", "deps", "usr")
+if !isdir(default_dir)
+    # Ensure default_dir exists
+    mkpath(default_dir)
+end
+
+immutable Environment
```
Comment (stevengj) [inline src/Conda.jl]: I'm not sure we want to declare a type that is nothing more than a wrapper around a pathname.

Comment (stevengj) [inline src/Conda.jl]: You could just do `typealias Environment Union{AbstractString,Symbol}`.   The only place you would actually need to convert the `Symbol` to a path would be in a `_set_conda_env(cmd, env::Symbol)` method.

File: deps/build.jl (hunk 7 above)
Comment (isuruf) [inline deps/build.jl]: I'm storing this in `deps.jl`. How is it different?

Comment (stevengj) [inline deps/build.jl]: Because when you re-run Pkg.build, you don't read deps.jl

Imagine where the user does `ENV["CONDA_JL_HOME"]="....."; Pkg.build("Conda")` and is happy with her Conda configuration.  Then, at some point later, in a separate Julia session, she does `Pkg.update()`, which updates Conda and internally calls `Pkg.build("Conda")`.  Since the `CONDA_JL_HOME` is no longer set, it forgets her setting and reverts to the default configuration.

Instead, the build script needs to remember `default_dir` from previous builds, and use it unless the `CONDA_JL_HOME` environment variable is explicitly set.

File: deps/build.jl [hunk 11]
```diff
@@ -0,0 +1,4 @@
...
+open("deps.jl", "w") do f
+    println(f, "default_dir=\"$(escape_string(default_dir))\"")
+end
```
Comment (stevengj) [inline deps/build.jl]: You should only write this file if has changed, in order to prevent Conda from being spuriously recompiled:
```jl
deps = "default_dir=\"$(escape_string(default_dir))\"\n"
if !isfile("deps.jl") || readstring("deps.jl") != deps
    write("deps.jl", deps)
end
```

File: deps/build.jl (hunk 7 above)
Comment (isuruf) [inline deps/build.jl]: Thanks. I didn't realize this.

File: deps/build.jl [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+elseif isfile("deps.jl")
+    include("deps.jl")
+else
+    default_dir = abspath(dirname(@__FILE__), "usr")
+end
+
+deps = "default_dir=\"$(escape_string(default_dir))\""
+
+if !isfile("deps.jl") || readchomp("deps.jl") != deps
+    open("deps.jl", "w") do f
```
Comment (stevengj) [inline deps/build.jl]: You can just do `write("deps.jl", deps)` rather than `open ... println ... end`, in which case you would include the newline in `deps` and use `readstring` rather than `readchomp`.

File: src/Conda.jl [hunk 13]
```diff
@@ -35,42 +35,98 @@ using Compat
 import Compat.String
 using JSON
 
-"Prefix for installation of all the packages."
-const PREFIX = abspath(dirname(@__FILE__), "..", "deps", "usr")
+deps_file = joinpath(dirname(@__FILE__), "..", "deps","deps.jl")
```
Comment (stevengj) [inline src/Conda.jl]: `const deps_file = ...`

File: src/Conda.jl [hunk 14]
```diff
@@ -35,42 +35,98 @@ using Compat
...
+
+if isfile(deps_file)
+    include(deps_file)
+else
+    error("Conda is not properly configured.  Run Pkg.build(\"Conda\") before importing the Conda module.")
+end
+
+typealias Environment Union{AbstractString,Symbol}
+
+RootEnv = default_dir
```
Comment (stevengj) [inline src/Conda.jl]: why not just use `default_dir` here?  Or call it `rootenv` in the `deps.jl` file?

(Note that CamelCase is used only for types in the typical Julia style.)

File: src/Conda.jl [hunk 15]
```diff
@@ -35,42 +35,98 @@ using Compat
...
+
+"Prefix for installation of the environment"
+function prefix(name::Symbol)
+    if (length(string(name)) == 0)
```
Comment (stevengj) [inline src/Conda.jl]: Don't use parentheses around the conditionals in an `if` statement.   Use `isempty(string(name))` rather than `length` (which is O(string length) thanks to Unicode).

File: src/Conda.jl [hunk 16]
```diff
@@ -35,42 +35,98 @@ using Compat
...
+        return error("Environment name should be non empty.")
```
Comment (stevengj) [inline src/Conda.jl]: `error` is not a return value.   Just call `error`.   Or `throw(ArgumentError("environment name must be non-empty"))` to have a more specific exception type.

File: src/Conda.jl [hunk 17]
```diff
@@ -35,42 +35,98 @@ using Compat
...
+    end
+    return joinpath(default_dir, "envs", string(name))
+end
+
+function prefix(path::Compat.ASCIIString)
+    if (!isdir(path))
```
Comment (stevengj) [inline src/Conda.jl]: no parens around condition

File: src/Conda.jl (hunk 15 above)
Comment (stevengj) [inline src/Conda.jl]: Since you call `string(name)` twice, it would be cleaner to define `sname = string(name)` once.

File: src/Conda.jl (hunk 4 above)
Comment (stevengj) [inline src/Conda.jl]: Define `conda_bat` = `joinpath(p, "conda.bat")`, rather than calling joinpath twice.

File: src/Conda.jl [hunk 18]
```diff
@@ -35,42 +35,98 @@ using Compat
...
+        return isfile(joinpath(p, "conda.bat")) ? joinpath(p, "conda.bat") : joinpath(p, "conda.exe")
+    else
+        return joinpath(script_dir(env), "conda")
+    end
+end
+const conda = conda_bin(RootEnv)
 
 "Path to the condarc file"
-const CONDARC = joinpath(PREFIX, "condarc-julia")
+const CONDARC = joinpath(prefix(RootEnv), "condarc-julia")
```
Comment (stevengj) [inline src/Conda.jl]: I thought `PREFIX == prefix(RootEnv)` already?

File: src/Conda.jl [hunk 19]
```diff
@@ -35,42 +35,98 @@ using Compat
...
     end
     for var in to_remove
-        pop!(env, var)
+        pop!(env_var, var)
     end
-    env["CONDARC"] = CONDARC
-    setenv(cmd, env)
+    env_var["CONDARC"] = joinpath(prefix(env), "condarc-julia")
+    env_var["CONDA_PREFIX"] = prefix(env)
+    env_var["CONDA_DEFAULT_ENV"] = prefix(env)
```
Comment (stevengj) [inline src/Conda.jl]: `env_var["CONDA_PREFIX"] = env_var["CONDA_DEFAULT_ENV"] = prefix(env)`

File: src/Conda.jl [hunk 20]
```diff
@@ -137,34 +191,40 @@ function _install_conda(force=false)
             end
         end
         Conda.add_channel("defaults")
+        # Update conda because conda 4.0 is needed and miniconda download installs only 3.9
+        run(_set_conda_env(`$(Conda.conda) update -y conda`))
+    end
+    if !isdir(prefix(env))
```
Comment (stevengj) [inline src/Conda.jl]: Shouldn't we throw an error in this case?

Comment (isuruf) [inline src/Conda.jl]: `prefix` already does if it is a full path. In case of a environment name, the environment has to be created here.

File: src/Conda.jl [hunk 21]
```diff
@@ -137,34 +191,40 @@ function _install_conda(force=false)
...
     end
 end
 
 "Install a new package."
-function add(pkg::AbstractString)
-    _install_conda()
-    run(_set_conda_env(`$conda install -y $pkg`))
+function add(pkg::AbstractString, env::Environment=RootEnv)
+    _install_conda(env)
+    run(_set_conda_env(`$(conda_bin(env)) install -y $pkg`, env))
```
Comment (stevengj) [inline src/Conda.jl]: It seems like we should just have a function
```jl
runconda(args::Cmd, env::Environment=RootEnv) =
    run(_set_conda_env(`$(conda_bin(env)) $args`, env))
```
and then we can just do
```
runconda(`install -y $pkg`, env)
```
etc. rather than re-typing `run(_set_conda_env($(conda_bin(env))  ...` several times

Comment (stevengj) [inline src/Conda.jl]: Note that this comment is still relevant.

File: deps/build.jl [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+
+if haskey(ENV, "CONDA_JL_HOME")
+    rootenv = ENV["CONDA_JL_HOME"]
+elseif isfile("deps.jl")
+    include("deps.jl")
+else
+    rootenv = abspath(dirname(@__FILE__), "usr")
+end
+
+deps = "rootenv=\"$(escape_string(rootenv))\""
```
Comment (stevengj) [inline deps/build.jl]: `const ROOTENV = ` (should be `const`, and since we are using all-caps for the other path constants we should be consistent here)

Comment (stevengj) [inline deps/build.jl]: should have `\n` at the end of the `deps` string

File: deps/build.jl [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+
+if !isfile("deps.jl") || Compat.readstring("deps.jl") != deps
```
Comment (stevengj) [inline deps/build.jl]: Just `readstring`.  No need for `Compat.`, since this function is exported.

Comment (isuruf) [inline deps/build.jl]: Doesn't work for me in Julia 0.4.6

Comment (isuruf) [inline deps/build.jl]: You are right. It is exported

File: src/Conda.jl (hunk 21 above)
Comment (isuruf) [inline src/Conda.jl]: There are a few `readstring(_set_conda_env(` as well. Any suggestions for a function name? or should I keep them as they are?

Comment (stevengj) [inline src/Conda.jl]: `readconda`?   Basically, anytime we have the same complicated function-call pattern appearing multiple times, it makes sense to put it in a new function.

File: src/Conda.jl [hunk 24]
```diff
@@ -137,34 +192,40 @@ function _install_conda(force=false)
...
 end
 
 "Update all installed packages."
-function update()
-    _install_conda()
+function update(env::Environment=rootenv)
+    _install_conda(env)
     for package in _installed_packages()
-        run(_set_conda_env(`$conda update -y $package`))
+        run(_set_conda_env(`$(conda_bin(env)) update -y $package`, env))
```
Comment (isuruf) [inline src/Conda.jl]: Shall I give all the packages as one command? Eg: `conda update -y numpy scipy`
Advantage would be that if conda cannot find a valid set of versions to update to, this will fail.

Comment (isuruf) [inline src/Conda.jl]: `conda update --all` is better

File: src/Conda.jl [hunk 25]
```diff
@@ -35,42 +35,110 @@ using Compat
...
+"Prefix for installation of the environment"
+function prefix(name::Symbol)
+    sname = string(name)
+    if isempty(sname)
+        throw(ArgumentError("Environment name should be non empty."))
+    end
+    return joinpath(ROOTENV, "envs", sname)
+end
+
+function prefix(path::Compat.ASCIIString)
```
Comment (stevengj) [inline src/Conda.jl]: Why not `AbstractString`?   We shouldn't limit ourselves to ASCII paths.

File: src/Conda.jl [hunk 26]
```diff
@@ -35,42 +35,110 @@ using Compat
...
+    if !isdir(path)
+        throw(ArgumentError("Path to conda environment is not valid."))
+    end
+    return path
+end
+
+if !isdir(ROOTENV)
+    # Ensure ROOTENV exists, otherwise defining constants below will throw errors
+    mkpath(ROOTENV)
+end
```
Comment (stevengj) [inline src/Conda.jl]: Should this be in an `__init__` function?   Otherwise it is only executed once, when Conda is precompiled, not every time Conda is imported.

Comment (stevengj) [inline src/Conda.jl]: Or should we throw an error if `!isdir(ROOTENV)`?   It seems like this should already exist after running `Pkg.build("Conda")`, no?

File: src/Conda.jl [hunk 27]
```diff
@@ -35,42 +35,110 @@ using Compat
...
+
+"Run conda command with environment variables set."
+function runconda(args::Cmd, env::Environment=ROOTENV)
+    _install_conda(env)
+    run(_set_conda_env(`$(conda_bin(env)) $args`, env))
+end
+
+"Run conda command with environment variables set and return the output as a string"
+function readconda(args::Cmd, env::Environment=ROOTENV)
```
Comment (stevengj) [inline src/Conda.jl]: Since this is only used with `JSON.parse`, maybe it should be called `parseconda` and be defined as
```jl
JSON.parse(readstring(_set_conda_env(`$(conda_bin(env)) $args --json`, env)))
```

File: src/bindeps_conda.jl [hunk 28]
```diff
@@ -1,57 +1,46 @@
 # This file contains the necessary ingredients to create a PackageManager for BinDeps
 using BinDeps
 
-type Manager <: BinDeps.PackageManager
+type ManagerType{T} <: BinDeps.PackageManager
     packages::Vector{Compat.ASCIIString}
```
Comment (stevengj) [inline src/bindeps_conda.jl]: Can package names be non-ASCII?  Probably we should use `Compat.UTF8String` here to be on the safe side.

File: src/bindeps_conda.jl [hunk 29]
```diff
@@ -1,57 +1,46 @@
...
 end
 
-function Base.show(io::IO, manager::Manager)
+"Manager for root environment"
+Manager = ManagerType{Symbol(PREFIX)}
+
+function Base.show{T}(io::IO, manager::ManagerType{T})
     write(io, "Conda packages: ", join(manager.packages, ", "))
```
Comment (stevengj) [inline src/bindeps_conda.jl]: Probably this should include `T` in the output.

File: src/bindeps_conda.jl [hunk 30]
```diff
@@ -1,57 +1,46 @@
...
 end
 
-BinDeps.can_use(::Type{Manager}) = true
+BinDeps.can_use{T}(::Type{ManagerType{T}}) = true
```
Comment (stevengj) [inline src/bindeps_conda.jl]: The type parameter is not necessary since you don't use it.  Just `BinDeps.can_use(::Type{ManagerType}) = true` suffices.

Comment (stevengj) [inline src/bindeps_conda.jl]: (Similarly for other functions below where you don't use `T`.)

File: src/bindeps_conda.jl [hunk 31]
```diff
@@ -1,57 +1,46 @@
...
 using BinDeps
 
-type Manager <: BinDeps.PackageManager
+type ManagerType{T} <: BinDeps.PackageManager
     packages::Vector{Compat.ASCIIString}
 end
 
-function Base.show(io::IO, manager::Manager)
+"Manager for root environment"
+Manager = ManagerType{Symbol(PREFIX)}
```
Comment (stevengj) [inline src/bindeps_conda.jl]: use `const` or `typealias`

File: src/bindeps_conda.jl [hunk 32]
```diff
@@ -1,57 +1,46 @@
 # This file contains the necessary ingredients to create a PackageManager for BinDeps
 using BinDeps
 
-type Manager <: BinDeps.PackageManager
+type ManagerType{T} <: BinDeps.PackageManager
```
Comment (stevengj) [inline src/bindeps_conda.jl]: `ManagerType` is confusing, since both `Manager` and `ManagerType` are types.  Maybe rename `ManagerType` to `EnvManager`?

File: test/runtests.jl [hunk 33]
```diff
@@ -3,82 +3,42 @@ using BinDeps
...
-    installed = Conda._installed_packages()
-
-    for (name, (_, package)) in PREVIOUS_PACKAGES
-        if !(name in installed)
-            Conda.add(package)
-        end
-    end
-end
-
 function main()
```
Comment (stevengj) [inline test/runtests.jl]: What's the point of having all of this in a function?  It's not typical for test scripts.

File: test/runtests.jl [hunk 34]
```diff
@@ -3,82 +3,42 @@ using BinDeps
...
-    Conda.add_channel("foo")
+    Conda.add_channel("foo", env)
 
-    Conda.rm_channel("foo")
-    channels = Conda.channels()
+    Conda.rm_channel("foo", env)
+    channels = Conda.channels(env)
     @test channels == ["defaults"]
-
-    teardown()
```
Comment (stevengj) [inline test/runtests.jl]: Should it delete the `test_conda_jl` environment afterwards?  And probably put the whole thing in a `try ... finally` block to guarantee the teardown even if the tests fail?

File: test/runtests.jl (hunk 33 above)
Comment (Luthaf) [inline test/runtests.jl]: It is not necessary, I used it to have nicer setup/teardown functions call. But it can go away if you prefer!

Comment (stevengj) [inline test/runtests.jl]: Since it is all really just one long script that is only executed once, there doesn't seem much point in putting it into a function.

File: src/Conda.jl (hunk 25 above)
Comment (isuruf) [inline src/Conda.jl]: Since we are downloading Anaconda2, the path  should be ASCII paths. See conda/conda#1180.
On the other hand, a user could download Anaconda3 and use a Python2 environment by setting `CONDA_JL_HOME`.

File: src/Conda.jl (hunk 26 above)
Comment (isuruf) [inline src/Conda.jl]: The logic is not there in `build.jl`. I can move this to `build.jl`

File: src/bindeps_conda.jl (hunk 28 above)
Comment (isuruf) [inline src/bindeps_conda.jl]: Yes, it can be. In Conda.jl `Compat.UTF8String`, `Compat.ASCIIString`, `AbstractString` and `Compat.String` are all used. Which should be the standard?

File: src/Conda.jl (hunk 26 above)
Comment (stevengj) [inline src/Conda.jl]: Thanks.

File: src/bindeps_conda.jl (hunk 28 above)
Comment (stevengj) [inline src/bindeps_conda.jl]: For function arguments, you want to use `AbstractString` so that any string type is accepted.  For `type` fields, you want to use a concrete type, and `Compat.UTF8String` (same as `String` in Julia 0.5) is the right choice because it can hold any Unicode string data.

File: src/bindeps_conda.jl (hunk 29 above)
Comment (isuruf) [inline src/bindeps_conda.jl]: We already have `Conda packages: symengine==0.2.0 at /home/isuru/.julia/v0.4/Conda/deps/usr/lib/libsymengine.so`. Printing `T` would print the path `/home/isuru/.julia/v0.4/Conda/deps/usr/` which is redundant.

Comment (isuruf) [inline src/bindeps_conda.jl]: @stevengj, ping.

**Comment by stevengj** [comment]
Looks good to merge, once tests are green.  Thnaks for sticking with this, @isuruf!
